File: 백준/Silver/2512. 예산/예산.py
import sys

input = sys.stdin.readline

n = int(input())
lst = list(map(int, input().split()))
m = int(input())

start, end = 0, max(lst)

while start <= end:
    sum = 0
    mid = (start + end) // 2

    for e in lst:
        if e > mid:
            sum += mid
        else:
            sum += e

    if sum > m:
        end = mid - 1
    else:
        start = mid + 1

print(end)

File: 백준/Silver/2512. 예산/예산.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        int[] lst = new int[N];
        String[] input = br.readLine().split(" ");

        int maxVal = 0;
        for (int i = 0; i < N; i++) {
            lst[i] = Integer.parseInt(input[i]);
            maxVal = Math.max(maxVal, lst[i]);
        }

        int M = Integer.parseInt(br.readLine());

        int start = 0, end = maxVal;
        while (start <= end) {
            int mid = (start + end) / 2;
            long sum = 0;

            for (int e : lst) {
                sum += Math.min(e, mid);
            }

            if (sum > M) {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }

        System.out.println(end);
    }
}
